From 87413a64ea226c82a0ef2b34b9c77543ded11230 Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Tue, 27 Nov 2018 17:04:10 +0800
Subject: [PATCH] usb: dwc_otg_310: pcd: don't set pullup twice

This patch adds a new flag pullups_connected to avoid
set pullup twice. It can help to fix the issue that
usb disconnect and reconnect again if we set pullup on
twice when usb connect to PC.

Change-Id: I756448cee9d2377695681d0394fada51a0c208f4
Signed-off-by: William Wu <william.wu@rock-chips.com>
---
 drivers/usb/dwc_otg_310/dwc_otg_pcd.h       | 2 ++
 drivers/usb/dwc_otg_310/dwc_otg_pcd_linux.c | 5 +++++
 2 files changed, 7 insertions(+)

diff --git a/drivers/usb/dwc_otg_310/dwc_otg_pcd.h b/drivers/usb/dwc_otg_310/dwc_otg_pcd.h
index 1c29390372f1..a6087f140a8c 100644
--- a/drivers/usb/dwc_otg_310/dwc_otg_pcd.h
+++ b/drivers/usb/dwc_otg_310/dwc_otg_pcd.h
@@ -251,6 +251,8 @@ struct dwc_otg_pcd {
 	struct cfiobject *cfi;
 #endif
 
+	/* true when pullup operation is set to on */
+	bool pullups_connected;
 	/** otg phy may be suspend in device mode, 1:suspend, 0:normal */
 	uint8_t phy_suspend;
 	/** vbus status in device mode */
diff --git a/drivers/usb/dwc_otg_310/dwc_otg_pcd_linux.c b/drivers/usb/dwc_otg_310/dwc_otg_pcd_linux.c
index 1d3ae00f9d00..36f273cdc6cc 100644
--- a/drivers/usb/dwc_otg_310/dwc_otg_pcd_linux.c
+++ b/drivers/usb/dwc_otg_310/dwc_otg_pcd_linux.c
@@ -744,13 +744,18 @@ static int dwc_otg_pcd_pullup(struct usb_gadget *_gadget, int is_on)
 		core_if = GET_CORE_IF(d->pcd);
 	}
 
+	if (is_on == pcd->pullups_connected)
+		return 0;
+
 	if (is_on) {
 		/* dwc_otg_pcd_pullup_enable(pcd); */
 		pcd->conn_en = 1;
 		pcd->conn_status = 0;
+		pcd->pullups_connected = true;
 	} else {
 		dwc_otg_pcd_pullup_disable(pcd);
 		pcd->conn_en = 0;
+		pcd->pullups_connected = false;
 	}
 
 	return 0;
-- 
2.35.3

